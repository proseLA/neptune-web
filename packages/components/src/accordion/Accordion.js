import PropTypes from 'prop-types';
import { useState } from 'react';

import AccordionItem from './AccordionItem';

const Accordion = ({ items, onClick, indexOpen }) => {
  const [itemsOpen, setItemsOpen] = useState(() =>
    items.map((value, index) => index === indexOpen),
  );
  const handleOnClick = (index) => {
    if (onClick) {
      onClick(index);
    }
    const newItems = [...itemsOpen];
    newItems[index] = !itemsOpen[index];
    setItemsOpen(newItems);
  };

  return items.map((item, index) => (
    <AccordionItem
      key={item.id || index}
      open={itemsOpen[index]}
      onClick={() => handleOnClick(index)}
      {...item}
    />
  ));
};

Accordion.propTypes = {
  indexOpen: PropTypes.number,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.node.isRequired,
      content: PropTypes.node.isRequired,
      icon: PropTypes.node,
    }),
  ).isRequired,
  onClick: PropTypes.func,
};

Accordion.defaultProps = {
  indexOpen: -1,
  onClick: null,
};

export default Accordion;
