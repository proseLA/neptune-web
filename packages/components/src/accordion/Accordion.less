@import (reference) '@transferwise/neptune-css/src/variables/neptune-tokens.less';
@import (reference) '@transferwise/neptune-css/src/less/mixins/_logical-properties.less';

.np-accordion-item {
  display: block;
  width: 100%;

  &--open {
    .media {
      border: 0 !important;
    }
  }

  .decision:hover {
    background-color: transparent;
  }

  &.np-accordion-item--dark {
    .media-body,
    .title-4 {
      color: @color-navy-content-primary;
    }
  }

  &__content {
    padding: 0 var(--size-12) var(--size-24) var(--size-12);
    border-bottom: 1px solid var(--color-border-neutral);

    @media (--screen-xs) {
      &.has-icon {
        .padding(left, 68px);
      }
    }
  }
}
