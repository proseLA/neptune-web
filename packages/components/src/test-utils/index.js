import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';

import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 *
 */
function customRender(
  ui,
  { locale = DEFAULT_LOCALE, messages = en, direction, ...renderOptions } = {},
) {
  // eslint-disable-next-line react/prop-types
  var Wrapper = ({ children }) => {
    return (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

/**
 * Custom `renderHook` function which wraps passed elements in Provider component
 * For more info: https://react-hooks-testing-library.com/usage/advanced-hooks#context
 *
 */
function customRenderHook(callback, { locale = DEFAULT_LOCALE, messages = en, direction } = {}) {
  return renderHook(callback, {
    // eslint-disable-next-line react/prop-types
    wrapper: ({ children }) => (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    ),
  });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render, customRenderHook as renderHook };
