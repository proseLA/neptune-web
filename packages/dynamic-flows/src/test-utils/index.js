import { Provider, translations as componentTranslations } from '@transferwise/components';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { FetcherProviderFromBaseUrl } from '../common/contexts/fetcherContext';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(
  component,
  locale = 'en-GB',
  // eslint-disable-next-line unicorn/no-object-as-default-parameter
  messages = { ...componentTranslations['en'] },
  baseUrl = 'https://test-url',
) {
  return mount(
    <Provider i18n={{ locale, messages }}>
      <FetcherProviderFromBaseUrl baseUrl={baseUrl}>{component}</FetcherProviderFromBaseUrl>
    </Provider>,
  );
}

function getI18n(
  locale = 'en-GB',
  // eslint-disable-next-line unicorn/no-object-as-default-parameter
  messages = { ...componentTranslations['en'] },
) {
  return { locale, messages };
}

export { mountWithProviders as mount, getMockFetchPromise, wait, getI18n };
