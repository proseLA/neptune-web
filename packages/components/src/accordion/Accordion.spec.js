import { Freeze, Emoji, New } from '@transferwise/icons';

import { render, screen, userEvent } from '../test-utils';

import Accordion from './Accordion';

describe('Accordion', () => {
  const items = [
    {
      title: 'Item 1 title',
      content: 'Item 1 content',
      icon: <Freeze />,
    },
    {
      title: 'Item 2 title',
      content: 'Item 2 content',
      icon: <New />,
    },
    {
      title: 'Item 3 title',
      content: 'Item 3 content',
      icon: <Emoji />,
    },
  ];

  describe('defaults', () => {
    it('renders an empty list when no items are passed', () => {
      expect(render(<Accordion items={[]} />).container).toMatchSnapshot();
    });
  });

  describe('items', () => {
    it('renders a list with all items closed if items passed', () => {
      render(<Accordion items={items} />);

      items.forEach((item) => {
        expect(screen.getByText(item.title)).toBeInTheDocument();
        expect(screen.queryByText(item.content)).not.toBeInTheDocument();
      });
    });

    it('renders a list with a specified item open if indexOpen is set', () => {
      render(<Accordion items={items} indexOpen={1} />);

      expect(screen.queryByText(items[0].content)).not.toBeInTheDocument();
      expect(screen.getByText(items[1].content)).toBeInTheDocument();
      expect(screen.queryByText(items[2].content)).not.toBeInTheDocument();
    });
  });

  describe('open and close', () => {
    it('can open and close an item when clicked', () => {
      const item = items[0];
      render(<Accordion items={items} />);

      userEvent.click(screen.getByRole('button', { name: item.title }));
      expect(screen.getByText(item.content)).toBeInTheDocument();

      userEvent.click(screen.getByRole('button', { name: item.title }));
      expect(screen.queryByText(item.content)).not.toBeInTheDocument();
    });
  });

  describe('onClick', () => {
    it('calls onClick when an item is clicked', () => {
      const onClick = jest.fn();

      render(<Accordion items={items} onClick={onClick} />);

      userEvent.click(screen.getAllByRole('button')[0]);

      expect(onClick).toHaveBeenCalledTimes(1);
      expect(onClick).toHaveBeenCalledWith(0);
    });
  });
});
