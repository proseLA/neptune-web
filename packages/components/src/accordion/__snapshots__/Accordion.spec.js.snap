// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion defaults renders an empty list when no items are passed 1`] = `<div />`;
